In [2]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

In [3]:
np.random.seed(0)
df = pd.read_csv("D:/Datasets/base_datos_2008.csv", nrows=1000000)
# df = df.sample(frac=1).head(100)

# Gestión de datos avanzados

## Correlaciones -  Entender las relaciones entre las variables

## Test de la Chi-CCuadrado

In [4]:
np.random.seed(0)
df= df[df["Origin"].isin(["HOU","ATL","IND"])] # vamos a saleccionar aquellos vuelos en los que el origen sea HOU ATL IND
df = df.sample(frac=1) # 
df = df[0:10000] # seleccionamos las primeras diez mil filas

In [5]:
df["BigDelay"] = df["ArrDelay"] > 30 # devuelve True si el retraso es mayor a 30 y false en caso contrario
# Creación de una tabla de contingencias
# con las columnas de estudio BigDelay y Origin. Margins = true lo que hace es sumar las filas y columnas
observados = pd.crosstab(index=df["BigDelay"], columns=df["Origin"], margins=True)
observados

Origin     ATL   HOU  IND    All
BigDelay                        
False     6818  1080  641   8539
True      1143   201  117   1461
All       7961  1281  758  10000

### Lectura de la tabla de contingencias
<p>
    6818 son los vuelos de ATL que se han retrazado menos de 30 y 1143 los vuelos con un retrazo mayor a 30
</p>

In [6]:
# importación del paquete que realiza el test
from scipy.stats import chi2_contingency
test = chi2_contingency(observados)
test

Chi2ContingencyResult(statistic=np.float64(2.0203704019430053), pvalue=np.float64(0.9178156315824011), dof=6, expected_freq=array([[ 6797.8979,  1093.8459,   647.2562,  8539.    ],
       [ 1163.1021,   187.1541,   110.7438,  1461.    ],
       [ 7961.    ,  1281.    ,   758.    , 10000.    ]]))

In [7]:
test[2]

6

In [8]:
esperados =pd.DataFrame(test[3])
esperados

0          1         2        3
0  6797.8979  1093.8459  647.2562   8539.0
1  1163.1021   187.1541  110.7438   1461.0
2  7961.0000  1281.0000  758.0000  10000.0

In [9]:
esperados_relativos = round(esperados.apply(lambda r: r/len(df) * 100 , axis=1),2 )
observados_relativos = round(observados.apply(lambda r : r/len(df) * 100 , axis=1),2)

In [10]:
observados_relativos

Origin      ATL    HOU   IND     All
BigDelay                            
False     68.18  10.80  6.41   85.39
True      11.43   2.01  1.17   14.61
All       79.61  12.81  7.58  100.00

In [11]:
esperados_relativos

0      1     2       3
0  67.98  10.94  6.47   85.39
1  11.63   1.87  1.11   14.61
2  79.61  12.81  7.58  100.00

### los margenes se han mantenido. esto tiene que ser asi de lo contrario algo estaria mal. el tes compara cada uno de los elementos. 

In [12]:
test[1]

np.float64(0.9178156315824011)

## JOINS. Trabajar con bases de datos relacionales

In [13]:
consumidores = [("A","Móvil"),
                ("B","Móvil"),
                ("A","Portátil"), 
                ("A","Tablet"),
                ("B","Tablet"),
                ("C","Portátil"),
                ("D","Smartwatch"),
                ("E","Consola")]

con_labels = ["Consumidor","Producto"]

con_df = pd.DataFrame. from_records(consumidores,columns = con_labels)

productores = [("a","Móvil"),
               ("a","Smartwatch"),
               ("a","Tablet"),
               ("b","Portátil"),
               ("c","Sobremesa"), 
               ("c","Portátil")]

prod_labels = ["Productor","Producto"]

prod_df = pd.DataFrame.from_records(productores,columns = prod_labels)

In [20]:
con_df

Consumidor    Producto
0          A       Móvil
1          B       Móvil
2          A    Portátil
3          A      Tablet
4          B      Tablet
5          C    Portátil
6          D  Smartwatch
7          E     Consola

In [21]:
prod_df

Productor    Producto
0         a       Móvil
1         a  Smartwatch
2         a      Tablet
3         b    Portátil
4         c   Sobremesa
5         c    Portátil